NotationV0/src/lib/runner.ts

```diff
@@ -23,4 +30,31 @@ export async function notationRunner(artifactRefs: string[], runCommand: (artifa
     if (failedArtifactRefs.length > 0) {
         throw new Error(taskLib.loc('FailedArtifacts', failedArtifactRefs.join(', ')));
     }
+
+}
+
+// WarningStream is a writable stream that extracts warnings from logs.
+class WarningStream extends Writable {
+  private buffer: string;
+
+  constructor() {
+    super();
+    this.buffer = '';
+  }
+
+  _write(chunk: Buffer, encoding: string, callback: () => void) {
+    this.buffer += chunk.toString();
+    const lines = this.buffer.split('\n');
+    this.buffer = lines.pop() || '';
+    // extract warnings from logs
+    for (const line of lines) {
+        if (line.startsWith('Warning:')) {
+            taskLib.warning(line);
+            taskLib.setTaskVariable(STATUS, WARNING);
+        } else {
+            console.log(line);
+        }
+    }
+    callback();
+  }
```

Comment:
Overall lgtm. One question though, why do we need this?

Reply:
Notation writes the warning message to STDOUT. Therefore, if the user doesn't open the build log, they will not see the warning. In this class, we extract the warning message and write it to the pipeline warning. Consequently, the user can see the warning message on the overview page of the pipeline.

Example:
![image](https://github.com/AzureCR/notation-ado-task/assets/43160897/216b7fa9-5463-466f-a6fd-87e9e6b13329)

Reply:
I see. Some thoughts on user readability:
1. Notation has a warning message looks like this: `Warning: This feature is experimental and may not be fully tested or completed and may be deprecated. Report any issues to \"https://github/notaryproject/notation`. This warning message might confuse user as a pipeline warning because user won't know which part is experimental. This warning message needs to be shown right after its command.
2. Since we support sign multiple refs, a ref array [`ref1:myTag`, `ref2:myTag`, `ref3:myTag`, ...] would trigger multiple warnings with the same content?

Reply:
1. I agree that for your warning example, we may not need to display it as a pipeline warning. However, for some warnings, users should know that they are genuine warnings, such as "Warning: Always sign the artifact using the digest (@sha256:...) rather than a tag (:v3)...". Displaying it on the overview page will assist users in resolving it.
Can we display the warning message if it is related to a security issue? This way, users will not receive unnecessary warnings.
2. Yes, users will see multiple warning messages with the same content. They can click on the warning to view the detailed logs of the task.

Reply:
Cool, sounds good to me.

Reply:
Updated to only extract "Warning: Always sign the artifact using the digest (@sha256:...) rather than a tag (:v3)..."